### PR TITLE
Models: fix type hints

### PR DESCRIPTION
**Pull request by adamjstewart**
Was stuck on a bus and got bored. This PR fixes all model-related type hints. There are now only ~500 remaining mypy warnings for the `uq_methods` directory to resolve.

As always, I'm barely aware of what the code is doing for these, so a close look during review would be appreciated.

### CODE REVIEW

File: lightning_uq_box/models/bnnlv/utils.py [hunk 1]
```diff
@@ -105,7 +105,7 @@ def retrieve_module_init_args(
     Returns:
         nn.Modules init args
     """
-    current_init_arg_names = list(inspect.signature(current_module.__init__).parameters)
+    current_init_arg_names = list(inspect.signature(current_module.__init__).parameters)  # type: ignore[misc]
```
Comment (adamjstewart) [inline lightning_uq_box/models/bnnlv/utils.py:108]: inspecting the function signature is not recommended, but I didn't know a better solution off the top of my head, so ignored this warning for now

File: lightning_uq_box/models/cards.py [hunk 2]
```diff
@@ -47,7 +47,7 @@ def forward(self, x: Tensor, t: Tensor) -> Tensor:
 class DiffusionSequential(nn.Sequential):
     """My Sequential to accept multiple inputs."""
 
-    def forward(self, input: Tensor, t: Tensor):
+    def forward(self, input: Tensor, t: Tensor) -> Tensor:  # type: ignore[override]
```
Comment (adamjstewart) [inline lightning_uq_box/models/cards.py:50]: `nn.Sequential` only accepts one input. We accept 2, breaking some expectations for subclasses, but otherwise harmless to ignore.

File: lightning_uq_box/models/fc_resnet.py [hunk 3]
```diff
@@ -52,7 +54,7 @@ def __init__(
             self.last = nn.Linear(features, num_outputs)
 
         if activation == "relu":
-            self.activation = F.relu
+            self.activation: Callable = F.relu
```
Comment (adamjstewart) [inline lightning_uq_box/models/fc_resnet.py:57]: Could probably make this more specific

File: lightning_uq_box/models/masked_ensemble/utils.py [hunk 4]
```diff
@@ -23,44 +23,41 @@ def convert_deterministic_to_masked_ensemble(
...
-                        padding=curr_layer.padding,
-                        dilation=curr_layer.dilation,
-                        groups=curr_layer.groups,
-                        bias=curr_layer.bias is not None,
+                        in_channels=value.in_channels,
+                        out_channels=value.out_channels,
+                        kernel_size=value.kernel_size,  # type: ignore[arg-type]
+                        stride=value.stride,  # type: ignore[arg-type]
+                        padding=value.padding,  # type: ignore[arg-type]
+                        dilation=value.dilation,  # type: ignore[arg-type]
```
Comment (adamjstewart) [inline lightning_uq_box/models/masked_ensemble/utils.py:44]: Wasn't quite sure what was going on here, so didn't know how to fix it other than ignore it

File: lightning_uq_box/models/prob_unet.py [hunk 5]
```diff
@@ -106,15 +108,15 @@ def __init__(
...
+        layers: list[nn.Module] = []
         for i in range(len(self.num_filters)):
             """
             Determine input_dim and output_dim of conv layers in this block.
             The first layer is input x output,
             All the subsequent layers are output x output.
             """
-            input_dim = self.input_channels if i == 0 else output_dim  # noqa: F821
             output_dim = num_filters[i]
+            input_dim = self.input_channels if i == 0 else output_dim
```
Comment (adamjstewart) [inline lightning_uq_box/models/prob_unet.py:119]: This was actually a bug, `output_dim` was not defined on line 116.

File: lightning_uq_box/models/prob_unet.py [hunk 6]
```diff
@@ -351,3 +354,5 @@ def forward(self, feature_map: Tensor, z: Tensor) -> Tensor:
             feature_map = torch.cat((feature_map, z), dim=self.channel_axis)
             output = self.layers(feature_map)
             return self.last_layer(output)
+        else:
+            return z
```
Comment (adamjstewart) [inline lightning_uq_box/models/prob_unet.py:358]: This might be wrong, but we have to return _something_ when `use_tile=False`

**Comment by adamjstewart** [comment]
No matter which failing tests I run, they always pass locally. Any idea why that might be the case?

File: lightning_uq_box/models/masked_ensemble/utils.py (hunk 4 above)
Comment (nilsleh) [reply lightning_uq_box/models/masked_ensemble/utils.py:44]: it is taking the parameters of a configured `torch.nn.Conv2d` layer and turning it into a `MaskedConv2d` layer. the whole function recursively tries to find all Conv2 layers in a configured pytorch architecture and replace them with the masked convolution layer

File: lightning_uq_box/models/masked_ensemble/utils.py [hunk 7]
```diff
@@ -23,44 +23,41 @@ def convert_deterministic_to_masked_ensemble(
...
-                deterministic_model._modules[name],
-                num_estimators=num_estimators,
-                scale=scale,
+                value, num_estimators=num_estimators, scale=scale
             )
-        elif "Conv2d" in deterministic_model._modules[name].__class__.__name__:
-            curr_layer = deterministic_model._modules[name]
-            if curr_layer.in_channels >= 10:
+        elif isinstance(value, nn.Conv2d):
+            if value.in_channels >= 10:
```
Comment (nilsleh) [inline lightning_uq_box/models/masked_ensemble/utils.py:32]: Also I cannot remember anymore why the docstring says 10 input channels. but that is wrong, since the method works also for inputs with less than 10 channels

**Comment by adamjstewart** [comment]
Feel free to push to this branch if you need to make any changes

File: lightning_uq_box/models/masked_ensemble/utils.py (hunk 7 above)
Comment (nilsleh) [reply lightning_uq_box/models/masked_ensemble/utils.py:32]: oh wait, maybe it's also correct, because the changes should only be applied to deeper layers. Will check the original repo again...

File: lightning_uq_box/models/masked_ensemble/utils.py (hunk 4 above)
Comment (adamjstewart) [reply lightning_uq_box/models/masked_ensemble/utils.py:44]: For some reason, the type hints for `nn.Conv2d` are less specific than the type hints for `MaskedConv2d` I guess

**Review by nilsleh — APPROVED** [review]

**Comment by adamjstewart** [comment]
There are now a lot of mypy errors that weren't there before:
```
lightning_uq_box/models/prob_unet.py:102: error: Cannot determine type of "output_dim"  [has-type]
lightning_uq_box/models/prob_unet.py:111: error: Argument 1 to "append" of "list" has incompatible type "Conv2d"; expected "AvgPool2d"  [arg-type]
lightning_uq_box/models/prob_unet.py:113: error: Argument 1 to "append" of "list" has incompatible type "ReLU"; expected "AvgPool2d"  [arg-type]
lightning_uq_box/models/prob_unet.py:117: error: Argument 1 to "append" of "list" has incompatible type "Conv2d"; expected "AvgPool2d"  [arg-type]
lightning_uq_box/models/prob_unet.py:121: error: Argument 1 to "append" of "list" has incompatible type "ReLU"; expected "AvgPool2d"  [arg-type]
lightning_uq_box/models/prob_unet.py:183: error: Argument 1 to "normal_" has incompatible type "Tensor | None"; expected "Tensor"  [arg-type]
lightning_uq_box/models/prob_unet.py:197: error: Incompatible types in assignment (expression has type "Tensor", variable has type "int")  [assignment]
lightning_uq_box/models/prob_unet.py:198: error: Incompatible types in assignment (expression has type "Tensor", variable has type "int")  [assignment]
lightning_uq_box/models/prob_unet.py:200: error: Incompatible types in assignment (expression has type "Tensor", variable has type "int")  [assignment]
lightning_uq_box/models/prob_unet.py:201: error: Incompatible types in assignment (expression has type "Tensor", variable has type "int")  [assignment]
lightning_uq_box/models/prob_unet.py:263: error: Argument 1 to "append" of "list" has incompatible type "ReLU"; expected "Conv2d"  [arg-type]
lightning_uq_box/models/prob_unet.py:267: error: Argument 1 to "append" of "list" has incompatible type "ReLU"; expected "Conv2d"  [arg-type]
lightning_uq_box/models/prob_unet.py:56: error: Argument 1 to "truncated_normal_" has incompatible type "Tensor | None"; expected "Tensor"  [arg-type]
tests/uq_methods/test_laplace.py:27: error: Invalid index type "str" for "DictConfig | ListConfig"; expected type "int | slice"  [index]
tests/uq_methods/test_laplace.py:28: error: Invalid index type "str" for "DictConfig | ListConfig"; expected type "int | slice"  [index]
```

**Comment by nilsleh** [comment]
Oh I now realize I messed up with the git branches... :/ 

Also can we add mypy to CI already to check the directories that have been checked, for example now mypy always checks the models in CI as well, and some day hopefully the rest?

**Comment by adamjstewart** [comment]
Yes, we can selectively enable mypy for parts of the repo (include or exclude certain modules). In fact, that's exactly what PyTorch does: https://github.com/pytorch/pytorch/blob/main/mypy.ini

**Comment by nilsleh** [comment]
Superseded by #293
